### DISCUSSION TITLE
Reason for relying on relic codes for UMD

### DISCUSSION THREAD

**Discussion by murugappanrm** [quasarframework/quasar#13030] [General - Components / Directives / etc] [answered]
@rstoenescu 

In one of the posts under issues, i was asked why " For the life of me" i was using relic codes and the post was locked even though it  still relates to Quasar issue and not (vuejs). Try looking in the Internet for a working example. You will get only relics. Well, for "the stressful life" of business solution developers, the reason why many of us revert to relics are these flaws with the software developers:

(1) Absence of adequate documentation
(2) No real working examples for the latest versions
(3) Outdated video tutorials in their own site which are themselves relics (recently there was a pull request)
(4) Releasing versions without proper impact assessment
(5) Failing to research on level of user base as compared to other frameworks
(6) Presenting usage options that they would not recommend.

I agree somewhat that Quasar is a good (not fantastic) product but unfortunately, using the current documentation, I assure Quasar will lose user base (new and migrating) if  it continues to live within those boundaries of existence as i have stated above.

**Comment by metalsadman** [comment]
Docs codepen examples are all in UMD.

**Reply by murugappanrm** [reply]
@metalsadman 

Thank you for your guidance. I too found that out a year ago. Those code in codepen are good and show us how to use the components. What i was looking for us an example of a full blown application with multiple pages and how to inject my own components (not Quasar components) into a "template" like page. In PHP we call them widgets or partials that could be loaded into a "template" page. 

There are many examples in YouTube and all of them were outdated  because they we using version 1 and vue 2 but none using UDM.  I hope you understand my predicament.

**Reply by smolinari** [reply]
@murugappanrm - Components are somewhat different than PHP templates. Yet, they are also in many ways similar. With components, you don't need partials. A component is injectable and/or consumable by other components. There is also the concept of slots in Vue, which is really powerful, but hard to grasp at first.  

And, if you really wish to use Vue (or Quasar) to get the most value out of it (them), you'd know not to use UMD at all. It's why most examples out there aren't about UMD usage, but rather SFC usage with a CLI. In fact, using UMD basically reduces Quasar to only a component library. If that is all you need, you might be better off with some other component library. But, you'll also find they too won't be offering much on UMD usage for the reasons I just mentioned. 

I'd also like to note that it is obvious you have a gap in knowledge about how Vue works overall. My suggestion to you is not to criticize anything, until you have a full understanding about it. Because, instead of helping, you end up wasting people's time and often get people frustrated. Just ask questions first. Most people are patient enough to answer even the simplest of questions. 

Scott

**Reply by murugappanrm** [reply]
@smolinari 

Thank you for your advice. You are right in that i do not have a deeper understanding of Vue.js. I come from the background of Assembler, C , Cobol, Fortran, Pascal etc. My extensive familiarity in modern languages is currently in HTML, CSS, Javascript. PHP and Python. With regards to Vue framework, i started on it only 3 months ago and perhaps i was wrong in thinking Quasar was so easy that people could "jump-start" in the usage of Vuejs. Well, at least that was i was made to believe from the documentation. I am so sorry in misunderstanding a product that is "only usable by experts". I realise and tested that Nuxt.js and Vuetify are two Vuejs based products that seem to be true to that principle, so far.

Again, from your feedback,  "... it is obvious you have a gap in knowledge about how Vue works overall" i gather that Quasar is only usable by people with superior or expert knowledge of Vue? Forgive me if that understanding is wrong. I guess criticisms and feedback are also "tabooed"? From my 40 years in IT development, I was taught that criticisms from any level should be analyzed, reviewed and used in the best possible ways to deliver the best of products. I guess I am wrong in my understanding of today's IT environment. I/we are a large group of business solution developers and we always used customer feedback and criticisms, no matter how small or trivial, to improve our products, especially in the medical field. From our experience, we found most often failure of our products stemmed from lack of or outdated documentation. It was stupid of me to offer what i thought was valuable criticism only to receive sarcasm and insults to someone who drives software development emotionally. Its funny, that Quasar gives user various usage options and then tell you "you'd know not to use UMD at all".  I fail to understand why for the  "For the life of me" (quote from rstoenescu) would they do this? Its stumps many of us who are following this post. Look at the video tutorial links provided by the site. Many of them were  dated in 2020 and outdated. So, i guess it would be wrong if i pointed this out as a feedback. Oh ok then, its their "funeral" if they want to continue that way. Writing software is not easy. It is a continuous evolutionary process of write, review, update and ensure. Failing in any one part of this is writing a death sentence for the software. At least that is what i learnt in my years in product delivery

To answer your question, why do we use UMD version? Simple, may of our apps were and are in Codeigniter CI4 and PHP and it will take unearthly costs to migrate them all to Quasar and it will be really stupid of us to do so. So believing that Quasar is an "easy to use product" and being fantastic, we thought of using the UMD option, learn as we migrate and later convert. I guess we were stupid in the thinking and strategy.  So that you may know, we could achieve all that we want by using CI4 which is simple to learn and powerful to use. We could do all that Quasar could do and much more without overloading with so many layers of extra fat (like Vuejs, a strategy that doomed IBM) and that too without a need for a long and tedious learning curve (as you stated, its only for expert in Vuejs). So your next question would be, why "On earth" would you want to use Quasar then? We wanted to deliver a better user experience and also avoid page refresh at the front-end at the quickest of possible time and also as an evolutionary process for UI components. Then we realised how stupid we were as  many developers are still using other layers for server side backend such as Laravel and python together with Vuejs to avoid the extra fat required. We have actually developed an entire api services for all our products using CI4 and integrated Vuejs (Angular and React  as well but dropped them as they were too highly proprietary) in a similar manner.

There were two people who have been humble and extremely helpful thus far - Danny and Luke Diebold. I get lots of constructive help from them. Like, in the one of the threads in the same post under issues, Luke was so help in producing a 1 hour plus video on how to install and use the vite features in just 2 days after the post. You know what, it works wonderfully but could not get it to work using the documentation.

**Reply by rstoenescu** [reply]
@murugappanrm  Nice background. My background is also in Assembler and C as I've been programming for embedded devices for lots of years before shifting my attention to web development exclusively.

Now getting back to the subject at hand: there's nothing emotional from me. It's just constructive criticism pointed back, but seems like you are generally used to having it one way only. The amount of energy that you spent here and in the other posts would have otherwise allowed you to get going with Vue (30 min) and Quasar (60 min) by reading both docs.

The point with UMD that Scott said was that "yeah, you could use UMD, but there's a way better path by using Quasar CLI" not that "UMD should not be used". As a side note, two of our bigger sponsors are using the UMD variant exclusively. The difference is that they are willing to read the docs, which for a newbie and seasoned professional alike wouldn't be considered outrageous. We do get and value constructive criticism, but when it's actually constructive and based on facts, like in a billion tickets that you can check out here. We wouldn't have gotten so far in 7 years if it weren't for listening to our devs' needs. But the main point is that they respectfully went through our requirements: read the docs, tried things out, then reported problems or ways to get things done even better. You know, the foundation of any two parties exchanging ideas. First we establish the basic rules of communication between us (like we clearly posting our requirements which you disregard completely) then acting upon them. If even defining a custom component in Vue is considered to be "expert Vuejs knowledge", then we're definitely not on the same page. Just to see the absurdity of the situation: my UMD (full page) example contained a custom component, and you asked about this exact thing while reposting my working example. Getting more absurd now: if you click on the "Codepen" button on all our examples in docs it opens up a new window which uses the UMD variant. And not the most absurd thing of all: 90% of the tickets raised here also have a Codepen attached highlighting the issue being reported -- so in UMD as well. Like almost everybody around here, including absolute beginners have figured it out. Except for of course, people who search the Internet rather than searching the docs itself (we do have a searchbar).

So, getting back to the reposting of my full page example which already contained the information you were asking: If you're not willing to put any effort into the conversation I fail to see how this going forward can be considered constructive by any standards.

I would also highly suggest reading about open-source and what it means for being a maintainer of such a project. About what it means to write and post your software for free and responding to each thread. It may be an eye-opener and you'd definitely appreciate even more the help that we (Scott, Luke and me -- all part of Quasar Team) have already offered you. But at this point you're killing the conversation because you're not actively participating as you're ignoring all of our advices and examples. So let's end it here like grown-ups.

**Reply by murugappanrm** [reply]
Hey @rstoenescu,

You started this with your sarcastic reply "For the life of me". 

Anyway, agreed. Lets close this. Like you said, i am wasting my time.

**Reply by smolinari** [reply]
@murugappanrm 
> We wanted to deliver a better user experience and also avoid page refresh at the front-end at the quickest of possible time and also as an evolutionary process for UI components

Um, with the UMD use of Vue (and of course, any component library built on Vue you choose to use too), you'd still be loading pages on the server i.e. needing page refreshes. That's the main point of wanting to even take on using the UMD version. And, if you do go all out and do all the work of handing over the interactivity to your Vue components, so the pages aren't refreshing on the server, then you'd need to turn your CI4 backend mainly into an API. And if you do accomplish all that effort, you might as well just go full blown SPA and take full advantage of Vue's and/ or Quasar's DX aka use the CLIs. So, yes, you could use Vue/ Quasar for the interactive parts of your MVC based architecture. It's just that it's not the norm, mainly because you lose out on the awesome DX I just mentioned and that is why you don't find a lot of examples on the Internet of using Vue or Quasar via UMD.  

Quasar and Vue are relatively simple to learn and use compared to Angular and React. And only Vue offers this UMD path of implementation. However, please don't confuse that path with "a simple way to convert", because it absolutely isn't. It's an architectural shift. And, any architectural shift will be painful. 

Scott

**Reply by murugappanrm** [reply]
Hi @smolinari ,

Well said sir. Yes in total agreement with you. That would be the long term strategy. I feel that our main problem and needs is being missed out. The main intention of using UMD is to provide a smooth deliver-while-learning path for the developers. I am sure you will agree, from the newbie perspective,  to learn the entire concept of the Vue and its components would take a certain amount of time. Especially, when i realized that Vue itself has made numerous evolutionary changes from Oct 2021 to March 2022. When i started learning, the CLI was version 4 and all the tutorials worked but midway the tutorials could not be used the same way as Vue CLI changed to 5.0.4 and Vite got injected. To "add pain to the sorrow" I struggled to find documentation on how to revert to Vue CLI version 4 again to continue to use the tutorials.

Anyway, to cut the chase, we wanted to use the UMD option, get the developers acquainted with Quasar components while grasping the full functionality of Quasar and Vue. Example, I am learning Vue Js full course and jumping ahead with Composition API because many the other add-ons like Pinia have all gone that route. That is why i was seeking example codes that will gives us a jump start on how to use the UMD and inject components. 

Allow me to give a detail explanation. In CI4, we do a layout of a general page, and then each controller model will inject their view code into the  layout. Very similar to Vue Js App.vue and its components. I agree the that the ultimate goal would be using Quasar CLI and i fully familiar with that, same applies to Nuxt and Vuetify as other alternatives. Our corporate websites are all built using Joomla and Wordpress which are PHP MVC platforms. We have integrated all our business apps into Joomla and many of them, including those based on CI4, are embedded or coupled into Joomla and page builders like Joomshaper.  In all cases we want our users to login into Joomla front-end CMS and then bind the user control information into our back-end apps, such as second level authentication and corporate info. We did create a Vue js app but had to interface to our CI4 API back-end server that delivers interfaces to MySQL, MSSQL, SQLite, Postgres databases and host of other services which would take too much works with pure JS platforms. Many of our these CI4 apps were using Bootstrap for UI and our intention was to replace that with Quasar components, hence the need for UMD to address the immediate needs. I am sure you would have seen the many tutorials on how to integrate Quasar into Laravel  (not replace), a PHP platform. To add on, not forgetting the language translations we have to provide in ASEAN which speaks English, Vietnamese, Cambodian, Thai, Malay, Tagalog etc. usinfg l18n component. 

I dont know if i have expressed my needs clearly but one thing for sure the need is not so simple as some have thought to be. This is kind of complexity we need to address as business solution developers.

**Reply by smolinari** [reply]
@murugappanrm 
> I am sure you will agree, from the newbie perspective, to learn the entire concept of the Vue and its components would take a certain amount of time.

It's not too bad. But, yes. You'd need some time. A day for the basics. A week or so for the basics of the ecosystem. 

It sounds to me like you are deeply locked in to the PHP world. I believe though, you can use Joomla and Wordpress as headless CMSs. And CI4 can also be used as an API framework too. So, I'd suggest taking the deeper dive (and the time) to go a full SFC structure with the CLI and develop any new secondary and additional SPA apps you need that way. Once you get the hang of developing with the CLI, you can think about completely replacing other sites/ apps with Vue. I'd personally avoid trying to migrate away from Bootstrap via UMD. Moving to Vue from Bootstrap is an architectural change, as I mentioned, and that won't work well with refactoring out Bootstrap in small chunks. You'll more than likely end up having to rewrite everything and all that effort will go much, much faster if you just go with the CLI version and start from scratch. Especially with Quasar.  

Scott

**Comment by rstoenescu** [comment, marked as answer]
You yourself mentioned that you don't have time to read the documentation ([here](https://github.com/quasarframework/vue-cli-plugin-quasar/issues/52#issuecomment-1080032821)). You and your team spent days searching the Internet for a solution that was right there in the docs rather than actually reading the current docs. In what world does it makes sense not to spend 5 minutes on the UMD docs page but rather to search the Internet for days and then to find a Quasar v0.17 page and example and using that with Quasar v2 instead? Is there anything more to say? I quote:

> So coming to your point on reading the material, do you think we have the time , not to mention the inconsistent terminologies used by each software developer? 

Then I serve you the exact page to read and a concrete example, just to see that you don't know actually know how to use Vue so the example makes no sense anyway. You seem to ask how to declare a Vue component even though the example that I gave you contains one. Just like you previously said that you followed the vite plugin guide to the letter just to find out later that it was far from the case.

Then you're complaining why I locked the discussion after completely hi-jacking a ticket related to one specific thing (vue-cli) with general topic discussions (first the vite plugin, then Quasar CLI, then UMD). Essentially, the ticket to which other people are subscribed (and so receive notifications) became "let's help murugappanrm with all their misunderstandings".

But aside from that, I spent energy writing compelling posts for you for free, but you seem to not understand Vue itself so naturally you turn to using big words like "Absence of adequate documentation" and "No real working examples for the latest versions" when you relate to Quasar. And I feel like there's a scent of entitlement on your side, which is not how open-source works. I mentioned that **if you know Vue**, then understanding Quasar comes really easy and all the examples and all the documentation suddenly makes all the sense in the world. We're not going to cover the abc of using Vue in our docs because it doesn't makes any sense. There's the Vue docs for that. Our target audience is for devs that already know Vue (or are willing to learn Vue from its source). We don't get bothered answering basic Vue questions, but we do get bothered if this turns into a habit and someone is using an open-source project as their learning ground. You don't owe us anything and we don't owe you anything, especially since you are not even a $1 donator. We also do get bothered by devs not reading the documentation for which thousands of hours were spent writing and improving just so that we don't have to reply to all the basic Quasar stuff over and over and over again. It's a sign of mutual respect to first read the actual docs and value our time too.

**Reply by murugappanrm** [reply]
Why are you getting emotional. You asked for the reasons and i gave you mine. They may not necessarily be right. Its for you to accept and review criticisms and make a choice to accept or reject. There is no need to get emotional and put words like " For the life of me". Things can be said in simple words. I was not upset or emotional. All i did was to reply your question. Thats it. Yes i am not a single $ donor but does mean we will not subscribe to it. Under no terms did i ever state the software was no good. My comments were on shortfalls. 

Thank you for all your assistance and good work. I will post your comments in the ASEAN IT User Group and let them decide. We are just in the process software evaluation for enterprise usage within ASEAN and we work very closely with NASCOM too. 

Lets close this topic.

**Comment by pdanpdan** [comment]
I think your tickets history on GitHub is pretty funny and shows some pattern:)

**Reply by murugappanrm** [reply]
I am sorry, i do not understand why its called a Ticket? I thought this was a discussion thread. Are they one and the same? I always thought tickets were issued for Problem Management. Oh well. terminology variations i guess. Hehehe..
